designs/CentrallyManagingNuGetPackageVersions-VSManagePackages.md

```diff
@@ -0,0 +1,95 @@
+3. When updating a centrally managed package version, the `Directory.Packages.props` file is updated, and all projects that use the package are restored with the new package version.
+4. When uninstalling/removing a package from a centrally managed project, the `PackageReference` is removed from the project file, but the `PackageVersion` is kept in the `Directory.Packages.props` file.
+
+## Non - Goals
+
+a. This spec is scoped to Visual Studio API interaction between components. Package Manager UI changes are out of scope. Command Line Interface is out of scope.
+b. Removing a `PackageVersion` from the `Directory.Packages.props` file is out of scope for the first version, but may be in scope in the future.
+
+## Scenarios
+
+The following scenarios will be enabled when the proposed design changes will be completed.
+
+1. Installing a package in a project with Package Manager UI or Package Manager Console.
+2. Installing a package from a quick action when editing a source file.
+3. Removing a package from a project with Package Manager UI or Package Manager Console.
+4. Updating the version of a package for all projects using that `Directory.Packages.props` file.
+
+## Requirements
+
+The requirements are defined in the [feature spec](https://github.com/NuGet/Home/wiki/Centrally-managing-NuGet-package-versions).
+
+Note that:
+
+* It's valid for some, not all, projects in the solution to use centrally managed package versions.
+* It's valid for different projects using central package versions to use different `Directory.Packages.props` files.
+* As is valid now, it's valid for some projects to use `packages.config`, while others use `PackageReference`. The `packages.config` projects cannot use centrally managed package versions, and individual `PackageReference` projects may opt in or out of central package version management.
+
+## Solution
+
+### Installing a package in a project
+
+Currently NuGet calls project system APIs to add a PackageReference to the project for the given package id and version. However, when the project uses centrally managed versions, the project file must not add the version on the `PackageReference` element.
+
+There are two categories of solutions.
+
+* API ignorant of centrally managed package versions
+
+Project system APIs would not be changed, but the implementation would need to be aware of how to detect if the feature is enabled, and know when to add a version to the `PackageReference` and when not to, and when to modify the `Directory.Packages.props` and when not not.
+
+There would also be a potential issue when a single user gesture causes a package to be installed into multiple projects, particularly when the `Directory.Packages.props` file does not already contain a `PackageVersion` for the package. If NuGet calls APIs to modify all the projects in parallel, there might be a race condition with trying to update `Directory.Packages.props`.
```

Comment:
In addition to Directory.Packages.props updates -we would need to consider that the project file and Directory.Packages.props needs to be updated inside of a transaction like model.

Reply:
It would be ideal if it was transation-like, but honestly I don't think it's important. The `Directory.Packages.props` needs to be modified first to add a version for the package, and once that's done the project file needs to be updated. If the project file fails to update, I don't think there's any real harm in the `Directory.Packages.props` file having a package version for a package that isn't referenced by any project.  Once we get agreement on which team will be writing to the props file, if they find it easy to make it transactional, great.

NuGet does a preview restore before "committing" (calling the project system APIs to modify the project). The preview restore determines which packages will be installed (finding all dependencies). By the time NuGet calls project system APIs, we already know that the restore graph is valid. At least in PMUI, I'm not so sure about PMC.